supervisor/monitor_solaris.go

```diff
@@ -1,38 +1,123 @@
+// +build solaris,cgo
+
 package supervisor
 
+/*
+#include <port.h>
+#include <poll.h>
+#include <stdio.h>
+#include <unistd.h>
+
+int portAssociate(int port, int fd) {
+	if (port_associate(port, PORT_SOURCE_FD, fd, POLLIN | POLLHUP, NULL) < 0) {
+		return 1;
+	}
+}
+
+int getFd(uintptr_t x) {
+	return *(int *)x;
+}
+*/
+import "C"
 import (
-	"errors"
+	"sync"
+	"unsafe"
 
+	"github.com/Sirupsen/logrus"
 	"github.com/docker/containerd/runtime"
 )
 
+// NewMonitor starts a new process monitor and returns it
 func NewMonitor() (*Monitor, error) {
-	return &Monitor{}, errors.New("Monitor NewMonitor() not implemented on Solaris")
+	m := &Monitor{
+		receivers: make(map[int]interface{}),
+		exits:     make(chan runtime.Process, 1024),
+		ooms:      make(chan string, 1024),
+	}
+	fd, err := C.port_create()
+	if err != nil {
+		return nil, err
+	}
+
+	m.epollFd = int(fd)
+	go m.start()
+	return m, nil
 }
 
+// Monitor represents a runtime.Process monitor
 type Monitor struct {
-	ooms chan string
+	m         sync.Mutex
+	receivers map[int]interface{}
+	exits     chan runtime.Process
+	ooms      chan string
+	epollFd   int
 }
 
+// Exits returns the channel used to notify of a process exit
 func (m *Monitor) Exits() chan runtime.Process {
-	return nil
+	return m.exits
 }
 
+// OOMs returns the channel used to notify of a container exit due to OOM
 func (m *Monitor) OOMs() chan string {
 	return m.ooms
 }
 
+// Monitor adds a process to the list of the one being monitored
 func (m *Monitor) Monitor(p runtime.Process) error {
-	return errors.New("Monitor Monitor() not implemented on Solaris")
+	m.m.Lock()
+	defer m.m.Unlock()
+	fd := p.ExitFD()
+	if _, err := C.port_associate(C.int(m.epollFd), C.PORT_SOURCE_FD, C.uintptr_t(fd), C.POLLIN|C.POLLHUP, unsafe.Pointer(&fd)); err != nil {
+		return err
+	}
+	EpollFdCounter.Inc(1)
+	m.receivers[fd] = p
+	return nil
 }
 
+// MonitorOOM adds a container to the list of the ones monitored for OOM
```

Comment:
If I'm not mistaken, there's no such thing on Solaris because it assumes that the application is well written and will handle a failed memory allocation accordingly.

Can a note be added to the doc specifying that there is no OOM-Killer  on Solaris and thus nothing to set up?